store/tikv/mock-tikv/mvcc_leveldb.go

```diff
@@ -395,10 +395,87 @@ func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64
 	return pairs
 }
 
-// ReverseScan implements the MVCCStore interface.
+// ReverseScan implements the MVCCStore interface. The search range is [startKey, endKey).
 func (mvcc *MVCCLevelDB) ReverseScan(startKey, endKey []byte, limit int, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
-	// TODO: Implement it.
-	panic("not implemented")
+	mvcc.mu.RLock()
+	defer mvcc.mu.RUnlock()
+
+	var mvccEnd []byte
+	if len(endKey) != 0 {
+		mvccEnd = mvccEncode(endKey, lockVer)
+	}
+	iter := mvcc.db.NewIterator(&util.Range{
+		Limit: mvccEnd,
+	}, nil)
+	defer iter.Release()
+
+	succ := iter.Last()
+	currKey, _, _ := mvccDecode(iter.Key())
+	helper := reverseScanHelper{
+		startTS:  startTS,
+		isoLevel: isoLevel,
+		currKey:  currKey,
+	}
+
+	for succ && len(helper.pairs) < limit {
+		key, ver, err := mvccDecode(iter.Key())
+		if err != nil {
+			break
+		}
+		if bytes.Compare(key, startKey) < 0 {
+			break
+		}
+
+		if bytes.Compare(key, helper.currKey) != 0 {
+			helper.finishEntry()
+			helper.currKey = key
+		}
+		if ver == lockVer {
+			var lock mvccLock
+			err = lock.UnmarshalBinary(iter.Value())
+			helper.entry.lock = &lock
+		} else {
+			var value mvccValue
+			err = value.UnmarshalBinary(iter.Value())
+			helper.entry.values = append(helper.entry.values, value)
+		}
+		succ = iter.Prev()
+	}
+	if len(helper.pairs) < limit {
+		helper.finishEntry()
+	}
+	return helper.pairs
+}
+
+type reverseScanHelper struct {
+	startTS  uint64
+	isoLevel kvrpcpb.IsolationLevel
+	currKey  []byte
+	entry    mvccEntry
+	pairs    []Pair
+}
+
+func (helper *reverseScanHelper) finishEntry() {
+	reverse(helper.entry.values)
+	helper.entry.key = NewMvccKey(helper.currKey)
+	val, err := helper.entry.Get(helper.startTS, helper.isoLevel)
+	if len(val) != 0 || err != nil {
```

Comment:
Is it possible that both `val` and `err` are nil?

Reply:
It can be a deleted entry.

Reply:
Aha, you're right.

Reply:
I just follow the old logic in mvcc.go, if both `val` and `err` are nil, this entry doesn't exist and it doesn't need to append them to results.